provider/plugin/plugin.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type PluginProvider struct {
+	client          *http.Client
+	remoteServerURL *url.URL
+}
+
+type PropertyValuesEqualsRequest struct {
+	Name     string `json:"name"`
+	Previous string `json:"previous"`
+	Current  string `json:"current"`
+}
+
+type PropertiesValuesEqualsResponse struct {
+	Equals bool `json:"equals"`
+}
+
+func NewPluginProvider(u string) (*PluginProvider, error) {
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		return nil, err
+	}
+	return &PluginProvider{
+		client:          &http.Client{},
+		remoteServerURL: parsedURL,
+	}, nil
+}
+
+// Records will make a GET call to remoteServerURL/records and return the results
+func (p PluginProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := p.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+	err = json.Unmarshal(b, &endpoints)
+	if err != nil {
+		return nil, err
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges will make a POST to remoteServerURL/records with the changes
+func (p PluginProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	u, err := url.JoinPath(p.remoteServerURL.String(), "records")
+	if err != nil {
+		return err
+	}
+	b, err := json.Marshal(changes)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", u, bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+	resp, err := p.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
```

Comment:
Could we have some optional debug-level logging here? 
I believe this will make it much easier to develop plugins, if developers can just run externalDNS with debug-level logging.

Reply:
This is probably also helpful on all other functions that hit an endpoint on the plugin provider service.

Reply:
This is a good idea yeah, I can add it.